Shopify/ruby-lsp
# Add semantic tokens for local arguments

### Motivation

Tracking the local environment enables a lot of features that we can't otherwise implement. For example, knowing all occurrences of a specific variable or keeping track of parameters that only exist inside a specific scope.

This RFC PR proposes a way in which we can keep track of the environment in a simple way that is consistent with our existing requests.

### Implementation

It's significantly easier to review per commit.
1. Create a new request parent class `RequestWithEnvironment` and `Support::Environment`. The `Support::Environment` class keeps track of the current environment for a scope and the `RequestWithEnvironment` automatically pushes and pops new environments as needed (as well as declaring local arguments)
2. Add support for semantic highlighting method parameters consistently inside their current scope

For reference, this is what local variables vs local arguments look like to me in this branch.
<img width="298" alt="Screen Shot 2022-09-02 at 1 38 30 PM" src="https://user-images.githubusercontent.com/18742907/188208301-8a204a5c-b722-4720-864e-2c3e3fe9b52a.png">


### Automated Tests

Added tests for the new type of request. Adapted and added more examples for semantic highlighting.

### Manual Tests

1. Fire up the LSP on this branch
2. Visit any file that has methods with arguments
3. Verify they are highlighted consistently

Note: depending on your theme configurations, there may actually not be any difference between highlighting of variables and parameters. This won't work for everybody, but here are my theme customizations to help
```json
{
  "editor.semanticTokenColorCustomizations": {
    "[YOUR_THEME_NAME_GOES_HERE]": {
      "rules": {
        "variable": {
          "foreground": "#3cc098",
        }
      }
    }
  },
  "editor.tokenColorCustomizations": {
    "[YOUR_THEME_NAME_GOES_HERE]": {
      "textMateRules": [
        {
          "scope": [
            "variable.parameter",
          ],
          "settings": {
            "foreground": "#ff9a3b",
          }
        },
      ]
    }
  }
}
```
